Watefall 1.0

TYWong99 opened the pull request:
- Edited `cli_test` to make it easier to use as a template
- Created `cli_waterfall` with general structure outlined
- Created `standalone/waterfall.py` with outline and pseudo-code included

Will be adding more descriptions as time goes.

TYWong99 committed:
Watefall 1.0

- Edited `cli_test` to make it easier to use as a template
- Created `cli_waterfall` with general structure outlined
- Created `standalone/waterfall.py` with outline and pseudo-code included

Will be adding more descriptions as time goes.

qtran122 commented on cli_waterfall.py:
```diff
@@ -0,0 +1,67 @@
+
+arg_description = 'Process a tiled level XML and generate object layers with waterfalls'
+arg_help1 = 'Name of the tiled level XML'
+arg_help2 = 'Controls the amount of information displayed to screen. 0 = nearly silent, 2 = verbose'
```
instead of the names "arg_help1" and "arg_help2" which are generic, let's have specific arg names. "arg_filename" and "arg_verbosity" for instance.

qtran122 commented on cli_waterfall.py:
```diff
@@ -0,0 +1,67 @@
+    log.SetVerbosityLevel(args.v)
+
+    # Use a playdo to read/process the XML
+    pattern_root = file_utils.GetPatternRoot()
```
unused? I'm guessing this is to get at the waterfall patterns later. So it's fine to keep. Put a comment for future, to convey your intention

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,129 @@
+#--------------------------------------------------#
+'''Public Functions'''
+
+def GetMetaDataFromBlockout(playdo):
```
I see what you're doing with this function, but it'd be more ideal to roll it into ScanForWaterfalls. No need to obtain this info, re-package it, and then unpackage it again a little later - just do it in one pass.

Especially since... (see below comment)

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,129 @@
+
+def GetMetaDataFromBlockout(playdo):
+	'''Obtain the tile2d and various meta data for the waterfall, based on tilelayer name
+	  e.g. _falls => sort_value is DEFAULT_SORT, color is DEFAULT_COLOR
+	       _falls_fg15 => sort_value is "fg_tiles,15", color is DEFAULT_COLOR
+	       _falls_bg_bbccff => sort_value is DEFAULT_SORT_BG, color is "#bbccff/1f"
```
(continued from above) the usage would NOT be like this.

The goal is to not have to set the sort layers or color manually. Still, it's good we did this since I was able to get into your headspace and see what you were thinking. I think the best way to convey what the tool needs to do is to get you an example input level XML. So I'll send that over soon.

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,129 @@
+
+
+def ScanForWaterFalls(list_meta_data):
+	'''...'''
+
```
definitely fill in descriptions (the really descriptive kind, describe the input args and return value). Because that's what's going to convey the intent for your design

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,129 @@
+
+
+def CreateWaterfalls(playdo, list_meta_data, list_waterfall_layer):
+	'''...'''
```
definitely fill in descriptions (the really descriptive kind, describe the input args and return value). Because that's what's going to convey the intent for your design

qtran122 reviewed (changes requested):
Good start. Some work to do. See comments 💪

TYWong99 committed:
Waterfall 1.1

Re-did much of the code in `waterfall.py` based on the new outline, plus inclusion of more descriptions.

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,188 @@
+OBJECTGROUP_PREFIX = "objects_waterfall_"
+
+LAYER_SORT_DIFF = 10	# Difference of sort_value between consecutive tilelayers
+LAYER_SORT_ADDON = 1	# Add to sort_value
```
No need to track this variable

TYWong99 replied:
You mean `OBJECTGROUP_PREFIX `? Or `LAYER_SORT_DIFF` and `ADDON`? Both of these are like "settings" IMO, where there's a small chance of being changed in the future. I think it's much safer to make them easily customisable here, unless you are absolutely sure that's never ever gonna happen.

qtran122 replied:
I am referring to just ADDON. Looking at it now, I see I had misinterpreted what this variable does? It has a value of 1 and is named ADDON, so I thought you'd be using it to add incrementally to the sort order. 
For example

curr_sort_num += LAYER_SORT_ADDON

Not sure what it's doing actually 🤔

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,188 @@
+	return sort_value
+	'''
+
+	return ""
```
See my trello comment

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,188 @@
+
+	Then return a list of tuples that contain all of the above informations
+	  Each tuple represents a new waterfall object layer to be added
+	'''
```
I find the return value is still not descriptive enough. I want this to be concrete - exactly what data is in each field. For example:

"Returns a list of tuples, where each tuple contains the data needed to create a waterfall. Each tuple contains the following data: (start_position, end_position, sort_layer, theme)

- start_position - a tuple of x,y coordinates signifying where the waterfall starts
- end_position - a tuple of x,y coordinates signifying where the waterfall ends
- sort_layer - a string specifying the sort layer and order of the waterfall. For example, "bg_tiles,15"
- theme - a string (for example, "lava" or "meadow"). This will be looked up later to set the waterfall's color, alpha, speed, and other effects

Feel free to copy/paste what I wrote into the actual description

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,188 @@
+
+	''' =====[Pseudo-code]=====
+
+	list_waterfall_layer = []
```
not a fan of the name "list_waterfall_layer", since what it is becomes confusing. Is it a layer? Is it a list?

Let's use the name "list_waterfalls_to_create"

TYWong99 replied:
How about just `list_all_waterfalls`? I kinda want to avoid starting a variable with `list_waterfall_something` since it makes searching (and replacing) slightly inconvenient.

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,188 @@
+
+
+def CreateWaterfalls(waterfall_template, playdo, list_waterfall_layer):
+	'''Read and convert each tuple in the list into a waterfalls object layer'''
+
```
I won't review this yet, since I'll just look at ScanWaterfalls for now. But the goal was to write descriptions for the functions, not write pseudocode. The description doesn't tell me what each of the args are:

TYWong99 committed:
Waterfall 1.2

Main changes:

- `ScanForWaterFalls()` mostly finished
- `CreateWaterfalls()` description expanded
- `_CountSortInLayer()` removed for being unused

qtran122 commented on logic/standalone/waterfall.py:
```diff
@@ -0,0 +1,190 @@
+'''Variables'''
+
+# Template settings
+OBJECT_SUFFIX_FG = " fg"
+OBJECT_SUFFIX_BG = " bg"
```
I'm kinda getting annoyed with these variables since it interrupts my train of thought to be reading code, see a variable I don't recognize and then scroll up to find what it is. Some uses cases of it are justified, but it's getting excessive IMO. Let's remove these 2 variables for now. In the future, I'll have the intern lint all the code files and get them up to a style I'd prefer (more on that later)

qtran122 reviewed (changes requested):
see trello comment

TYWong99 committed:
Waterfall 1.3

Main edits:

 - `standalone/aligner.py` missing variable re-added
 - `common/level_playdo.py` function added to get all tiles2d
 - `standalone/waterfall.py`
  - Requested changes included
  - Can now scan multiple layers with same name
  - Can now grab the necessary info from tiles2d into list of waterfalls
  - Added toggle for debug printing function

qtran122 committed:
Merge branch 'main' into Waterfall
